sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -238,15 +188,17 @@ static AZ_NODISCARD az_result az_token_credential_add_token_header(
   AZ_RETURN_IF_FAILED(az_mut_span_copy(entire_buf, AZ_STR("Bearer "), &bearer));
   post_bearer = az_mut_span_drop(entire_buf, bearer.size);
 
-  az_span_builder auth_url_builder = az_span_builder_create(post_bearer);
+  uint8_t auth_url_buf[100] = { 0 };
+  az_span_builder auth_url_builder
+      = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(auth_url_buf));
+
   AZ_RETURN_IF_FAILED(
       az_token_credential_get_resource_url(az_mut_span_to_span(hrb->url), &auth_url_builder));
 
   az_span const auth_url = az_span_builder_result(&auth_url_builder);
 
   az_span token = { 0 };
-  AZ_RETURN_IF_FAILED(az_token_credential_get_token(
-      credential, auth_url, az_mut_span_drop(post_bearer, auth_url.size), &token));
+  AZ_RETURN_IF_FAILED(az_token_credential_get_token(credential, auth_url, post_bearer, &token));
 
   az_mut_span unused;
   AZ_RETURN_IF_FAILED(az_mut_span_move(post_bearer, token, &unused));
```

Comment:
I would like to review and maybe propose a different design for this entire method `az_token_credential_add_token_header`

This is what is currently do in summary:
1)  Gets a span from credential->token buffer (note: we force this buffer to be large enough to hold an entire HTTP_RESPONSE with a token)
2)  Adding the string `Bearer ` to buffer
3)  Calculate a new span after `Bearer ` to continue using crendential->token
4)  Allocate new buffer in stack to hold auth_url
5)  Calculate resource url and write it to previous buffer
6)  Get Token:
    6.1)  Allocate new buffer in stack to hold auth request
    6.2)  Build URL
    6.3)  Allocate new buffer in stack for body request
    6.4)  Build body and create a span pointing to where the secret is written
    6.5)  Create pipeline and send request using  span after `Bearer ` to hold http_response back
    6.6)  Clean body request buffer (step 6.3) (only the secret part).  // we could just clean it all or let it be gone after function return
  7) parse span after `Bearer ` holding http response to get body
  8) create new span to get token from body 
  9) Move credential->token memory pointed by span after `Bearer ` to keep only the token
  10) Finally add a new header to request with token!

Things to consider ->  2 buffers in stack created and amazing credential-token buffer re-usability
                               ->  Memory move required

What I would like to propose:
1)  Allocate new buffer in stack to hold auth_url
2)  Calculate resource url and write it to previous buffer
3)  Allocate new buffer in stack to hold a Token only  
4)  Get Token:
    4.1)  Allocate new buffer in stack to hold auth request url
    4.2)  Allocate new buffer in stack to hold auth request body
    4.3)  Allocate new buffer in stack to hold auth response
    4.4)  Build URL
    4.5)  Build Body
    4.6)  Create pipeline and send request. Get response in auth response
    4.7)  Parse http_response to get body
    4.8)  Get token from body and write it into Token only span (passed from previous function call)
5)  Create builder for Credential->Token
6)  Write `Bearer `
7)  Write Token only
8)  Finally add a new header to request with token!

So, yes, more buffers in stack!  but now Credential-> Token can be a little smaller, because it doesn't need to hold a one point a huge http_response (with all headers and entire body)

let me know your comments

Reply:
Thank you for a large feedback, I'll look into it and let you know in the further comments.
Regarding cleaning up the secret on the stack: yes, I think we need it, because 
```
get_token()
{
    char secret buf[] = ...
}

steal_secret()
{
    char restore_secret;
    printf(restore_secret);
}

demo()
{
    get_token();
    steal_secret();
}
```

Reply:
ok ok. security comes first. 👍   tks for the explanation. 
I would go and clean the entire buffer then (not just secret) so we don't provide any info like secret length =P... And we also protect even data like tenant_id

Reply:
Thank you Victor, I applied the principles from your longest response, and I like the code much better now!